Test and automatic Update default package version

wli3 opened the pull request:
partial https://github.com/dotnet/sdk/issues/2275

-  Add test to CLI that will catch portable not move forward that is similar to dotnet/cli#8919
-   CLI sets UseBundledVersionAsDefaultVersion based on patch version == 0 and is preview == true

wli3 commented:
@dotnet-bot test this

wli3 commented:
@dotnet-bot test this please

wli3 commented:
@dotnet-bot test this please

wli3 commented on test/EndToEnd/GivenPortableApps.cs:
```diff
@@ -0,0 +1,87 @@
+    {
+        [Theory]
+        [ClassData(typeof(SupportedNetCoreAppVersions))]
+        public void ItDoesNotRollsForwardToTheLatestVersion(string minorVersion)
```
@dsplaisted does this test looks good? And I need to keep work on Generate the flag `UseBundledVersionAsDefaultVersion` next

dsplaisted replied:
Grammar nit: rename the method to `ItDoesNotRollForwardToTheLatestVersion` ("roll" instead of "rolls").

wli3 commented on test/EndToEnd/GivenSelfContainedAppsRollForward.cs:
```diff
@@ -13,13 +12,13 @@
+    public partial class GivenSelfContainedAppsRollForward : TestBase
     {
 
-        [Theory(Skip = "Runtime 1.1 support for openSUSE and Fedora 27 needed")]
```
@dsplaisted new Linux lags are failing. At the same time, it is failing on assertion `floatedVersion` is 2.0.6 instead of 2.0.7. This is strange, it means 2.0.7 is not in the feed. Do you know why? It should be in myget at least

wli3 replied:
@johnbeisner is the cause of https://github.com/dotnet/cli/pull/9202 fix? Can I reenable it?

wli3 replied:
@dsplaisted that caught an actual bug

https://github.com/dotnet/cli/blob/508f902ed0d49891400a806e9b9785e5434d9789/build/MSBuildExtensions.targets#L174 will write out the following to disk. Which equal to no op. Not good...

```xml
    <LatestPatchVersionForNetCore1_0 Condition="'1.0.11' == ''">1.0.11</LatestPatchVersionForNetCore1_0>
    <LatestPatchVersionForNetCore1_1 Condition="'1.1.8' == ''">1.1.8</LatestPatchVersionForNetCore1_1>
    <LatestPatchVersionForNetCore2_0 Condition="'2.0.7' == ''">2.0.7</LatestPatchVersionForNetCore2_0>
```

dsplaisted replied:
Nice catch.  Looks like the `$` needs to be escaped in MSBuildExtensions.targets.

wli3 replied:
that is the end result now. Looks alright
```xml
<!--
***********************************************************************************************
Microsoft.NETCoreSdk.BundledVersions.props

WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
          created a backup copy.  Incorrect changes to this file will make it
          impossible to load or build your projects from the command-line or the IDE.

Copyright (c) .NET Foundation. All rights reserved.
***********************************************************************************************
-->
<Project>
  <PropertyGroup>
    <BundledNETCoreAppTargetFrameworkVersion>2.1</BundledNETCoreAppTargetFrameworkVersion>
    <BundledNETCoreAppPackageVersion>2.1.1-servicing-26605-02</BundledNETCoreAppPackageVersion>
    <BundledNETStandardTargetFrameworkVersion>2.0</BundledNETStandardTargetFrameworkVersion>
    <BundledNETStandardPackageVersion>2.0.3</BundledNETStandardPackageVersion>
    <BundledNETCorePlatformsPackageVersion>2.1.0</BundledNETCorePlatformsPackageVersion>
    <BundledAspNetCoreAllTargetFrameworkVersion>2.1</BundledAspNetCoreAllTargetFrameworkVersion>
    <BundledAspNetCoreAllPackageVersion>2.1.1-rtm-30818</BundledAspNetCoreAllPackageVersion>
    <BundledAspNetCoreAppTargetFrameworkVersion>2.1</BundledAspNetCoreAppTargetFrameworkVersion>
    <BundledAspNetCoreAppPackageVersion>2.1.1-rtm-30818</BundledAspNetCoreAppPackageVersion>
    <NETCoreSdkVersion>2.1.400-preview-009010</NETCoreSdkVersion>
    <_NETCoreSdkIsPreview>true</_NETCoreSdkIsPreview>

    <!-- Default patch versions for each minor version of ASP.NET Core -->
    <DefaultPatchVersionForAspNetCoreAll2_1>2.1.1-rtm-30818</DefaultPatchVersionForAspNetCoreAll2_1>
    <DefaultPatchVersionForAspNetCoreApp2_1>2.1.1-rtm-30818</DefaultPatchVersionForAspNetCoreApp2_1>
    <UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion>false</UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion>

    <!-- Latest patch versions for each minor version of .NET Core -->
    <LatestPatchVersionForNetCore1_0 Condition="'$(LatestPatchVersionForNetCore1_0)' == ''">1.0.11</LatestPatchVersionForNetCore1_0>
    <LatestPatchVersionForNetCore1_1 Condition="'$(LatestPatchVersionForNetCore1_1)' == ''">1.1.8</LatestPatchVersionForNetCore1_1>
    <LatestPatchVersionForNetCore2_0 Condition="'$(LatestPatchVersionForNetCore2_0)' == ''">2.0.7</LatestPatchVersionForNetCore2_0>
  </PropertyGroup>
</Project>

```

wli3 commented on build/MSBuildExtensions.targets:
```diff
@@ -132,6 +132,13 @@
+    <GetUseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion
+      BundledNETCoreAppPackageVersion="$(BundledNETCoreAppPackageVersion)">
+      <Output TaskParameter="UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion"
+              PropertyName="_UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion" />
```
@dsplaisted is the property name good?

dsplaisted replied:
It's pretty long, but I think it's OK.

wli3 commented on build_projects/dotnet-cli-build/GetUseBundledVersionAsDefaultVersion.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+            var parsedVersion = NuGetVersion.Parse(BundledNETCoreAppPackageVersion);
+            UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion =
+                (parsedVersion.Patch == 0) && parsedVersion.IsPrerelease ? "true" : "false";
```
use string "true" and "false" to avoid "True" and "False"

wli3 commented:
Due to https://github.com/dotnet/cli/issues/9512 I cannot add unit test for the version. I created a separate project to do sanity check. Please make sure it is expected

```csharp
        [Theory]
        [InlineData("2.1.0-preview1-26116-04", "true")]
        [InlineData("2.1.0", "false")]
        [InlineData("2.1.1", "false")]
        [InlineData("2.1.1-preview1-26116-04", "false")]
        public void Test1(string version, string result)
        {
            var task = new GetUseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion();

            task.BundledNETCoreAppPackageVersion = version;
            task.Execute();
            task.UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion.Should().Be(result);
        }
    }
```

dsplaisted commented on build/MSBuildExtensions.targets:
```diff
@@ -132,6 +132,13 @@
     </PropertyGroup>
 
+
+    <GetUseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion
```
Add a comment here explaining what this task is doing (setting the property to true if patch == 0 and preview == true), and why it is doing it (so that we don't need to manually update the version selection logic between when we ship a final release and when we ship the first patch).

wli3 replied:
Fixed according to the suggestion, and I updated the comment

dsplaisted commented on build/MSBuildExtensions.targets:
```diff
@@ -180,6 +187,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <!-- Default patch versions for each minor version of ASP.NET Core -->
     <DefaultPatchVersionForAspNetCoreAll2_1>$(_DefaultPatchVersionForAspNetCoreAll2_1)</DefaultPatchVersionForAspNetCoreAll2_1>
     <DefaultPatchVersionForAspNetCoreApp2_1>$(_DefaultPatchVersionForAspNetCoreApp2_1)</DefaultPatchVersionForAspNetCoreApp2_1>
+    <UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion>$(_UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion)</UseBundledNETCoreAppPackageVersionAsDefaultNetCorePatchVersion>
```
Move this out of the "section" for ASP.NET Core.  I'd put it right after BundledNETCoreAppPackageVersion is declared.

wli3 replied:
Fixed as suggested

dsplaisted commented on test/EndToEnd/GivenPortableApps.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace EndToEnd
+{
+    public class GivenPortableApps : TestBase
```
The term our docs use is "Framework dependent", so use that instead of "Portable" here.

wli3 replied:
renamed

dsplaisted commented on test/EndToEnd/GivenPortableApps.cs:
```diff
@@ -0,0 +1,87 @@
+
+            project.Save(projectPath);
+
+            //  Get the version rolled forward to
```
It's not actually supposed to roll forward, so change this comment to "Get the resolved version of .NET Core" or similar

wli3 replied:
changed to suggested

dsplaisted commented on test/EndToEnd/GivenPortableApps.cs:
```diff
@@ -0,0 +1,87 @@
+            var versionInAssertsJson = GetNetCoreAppVersion(assetsFile);
+            versionInAssertsJson.Should().NotBeNull();
+
+            if (versionInAssertsJson.IsPrerelease)
```
For this test, I think the condition here should be:

```c#
if (versionInAssertsJson.IsPrerelease && versionInAssetsJson.Patch == 0)
```

This ensures that if the bundled version is, for example, a prerelease of .NET Core 2.1.1, that we don't roll forward to that prerelease version for framework-dependent deployments.

wli3 replied:
Fixed and updated the comment

dsplaisted commented on test/EndToEnd/GivenSelfContainedAppsRollForward.cs:
```diff
@@ -118,25 +117,11 @@ public void WeCoverLatestNetCoreAppRollForward()
 
-                SupportedNetCoreAppVersions.Select(v => $"netcoreapp{v[0]}")
+                SupportedNetCoreAppVersions.Versions.Select(v => $"netcoreapp{v[0]}")
                     .Should().Contain(targetFramework, $"the {nameof(SupportedNetCoreAppVersions)} property should include the default version " +
```
Change "property" to "class" in this message now that you've refactored it (or do `{nameof(SupportedNetCoreAppVersions)}.{nameof(SupportedNetCoreAppVersions.Versions)}`)

dsplaisted commented:
> Due to #9512 I cannot add unit test for the version. I created a separate project to do sanity check. Please make sure it is expected

@wli3 I'm not following why we can't have test coverage here.  We should probably discuss in person.

wli3 commented:
@dotnet-bot test Debian8.2 x64 Debug Build

wli3 commented:
@dsplaisted finally fixed the problem. The problem is not RID, it is nuget will use whatever response first. So according to which response first, the latest runtime will change. I added https://github.com/dotnet/cli/pull/9507/commits/5e43e000f90b1757e14c20cefaac1c931e50e12d to use only myget dotnet-core feed. And I updated 1.0.X and 1.1.X version according to the latest in the feed https://github.com/dotnet/cli/pull/9507/commits/103e1878853c65680140f0bc1bcfe96b41033187
